Command group test

petervdonovan opened the pull request.

Bobman31 committed:
Added Lift System Command Group

Bobman31 committed:
Merge branch 'master' into CommandGroupTest

Bobman31 committed:
Revert "Merge branch 'master' into CommandGroupTest"

This reverts commit a9736867ed69ebaa8af242b90486d936b08c005e, reversing
changes made to dc0936f40d7a44c44103d54678056b1615534f9c.

MelissaGWhite committed:
Fixes

Fixed some issues so the code can deploy.

MelissaGWhite committed:
Working Drive Code

MelissaGWhite committed:
Update to make lift work

Bobman31 committed:
Added Arm Angle

MelissaGWhite committed:
Fixed Arm Angle

MelissaGWhite committed:
Update WheelArm.java

petervdonovan committed:
Update ArmAngle.java

petervdonovan committed:
Update RobotMap.java

petervdonovan committed:
Update LiftSystem.java

petervdonovan committed:
Update WheelArm.java

niclop03 committed:
Attempt to fix arm angle

niclop03 committed:
Merge branch 'CommandGroupTest' of https://github.com/poly-rabbotics/2019-Deep-Space into CommandGroupTest

MelissaGWhite committed:
Fixed Port Numbers

MelissaGWhite committed:
More changes

Bobman31 committed:
Create ExampleCommandGroup.java

Bobman31 committed:
Update ExampleCommandGroup.java

Bobman31 committed:
Talons -> VictorSP

MelissaGWhite committed:
Update RobotMap.java

MelissaGWhite committed:
Temporary resolving conflicts

MelissaGWhite committed:
fixed lift discrepencies (with solenoids and names)

MelissaGWhite committed:
Added a bracket

MelissaGWhite committed:
Merge branch 'CommandGroupTest' of https://github.com/poly-rabbotics/2019-Deep-Space into CommandGroupTest

Bobman31 committed:
Fixing Arm Angle (and Juliana's psyche)

MelissaGWhite committed:
Revert "Merge branch 'CommandGroupTest' of https://github.com/poly-rabbotics/2019-Deep-Space into CommandGroupTest"

This reverts commit a6207a8a972ef4932bfbe7515875fa56079f5dae, reversing
changes made to 4a6aac979c6fe0e8d355610dba484a6a5c7f4dd3.

MelissaGWhite committed:
Fixed Arm Anlge (and will's Psyche)

MelissaGWhite committed:
Update RobotMap.java

MelissaGWhite committed:
added code for cameraserver to view lifecams ...

niclop03 committed:
Updates Friday Before Bag Day

niclop03 committed:
Merge branch 'CommandGroupTest' of https://github.com/poly-rabbotics/2019-Deep-Space into CommandGroupTest

niclop03 committed:
Added Cameras

MelissaGWhite committed:
Update RobotMap.java

MelissaGWhite committed:
Fixed port numbers for VictorSPs

markpowers commented:
I agreed today we should do this and then do much smaller bites of code going forward. We still need lots of comments and/or readme content on these files. @Bobman31

jordan-powers commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -44,7 +57,26 @@
+    hatchCamera = CameraServer.getInstance().startAutomaticCapture("front",0);
+    hatchCamera.setResolution(640, 480);
+    portCamera = CameraServer.getInstance().startAutomaticCapture("back",1);
+    portCamera.setResolution(1280, 960);
```
Have you tested these resolutions? Don't forget the bandwidth limit is 4 Mbits/second

jordan-powers commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -44,7 +57,26 @@
+    portCamera = CameraServer.getInstance().startAutomaticCapture("back",1);
+    portCamera.setResolution(1280, 960);
+    
+    // new Thread(()->{
```
You DEFINITELY want to define a new class for this if you end up using it. Don't just put everyting in Robot.java, that'll get really confusing really fast

jordan-powers commented on src/main/java/frc/robot/commands/ArmAngleCommand.java:
```diff
@@ -23,6 +24,19 @@ protected void initialize() {
   // Called repeatedly when this Command is scheduled to run
   @Override
   protected void execute() {
+    DriveController controller = Robot.controller;
```
This WONT WORK!!
To show why, let's walk through a couple of loops of the code

**Loop 1**
_No buttons are pressed, the arms are not moving_
Do the arms think they're moving? no
Are any buttons pressed? no
_Nothing is written to the motor_

**Loop 2**
_User has pushed a button, but has not had time to let go yet. The arms are not moving_
Do the arms think they're moving? no
Are there any buttons pressed? yes, tell the motor to move
_The motors are told to move_

**Loop 3**
_User is still holding down the button. The arms have been told to move, but haven't started moving yet because of inertia_
Do the arms think they're moving? Yes
Are any buttons being pressed? Yes, but the button press was handled in the previous loop so this evaluates to No
_Nothing is written to the motor, so they continue at the previous set speed_

**Loop 4**
_User is still holding down the button. The arms are moving_
Do the arms think they're moving? Yes
Are any buttons being pressed? Yes, but the button press was handled in the previous loop so this evaluates to No
_Nothing is written to the motor, so they continue at the previous set speed_

**Loop 5**
_User has released the button. The arms are moving_
Do the arms think they're moving? Yes
Are any buttons being pressed? Yes, but the button press was handled in the previous loop so this evaluates to No
_Nothing is written to the motor. The motor notices it's been on for a while, but hasn't been updated. It thinks the code has frozen or stopped, throws an "Motor not updated enough exception", and stops the motor_

I thought I had made this clear, but apparently not.
**ALWAYS UPDATE THE MOTOR IN EVERY LOOP OF THE COMMAND**
In case you didn't get the message
**ALWAYS, ALWAYS UPDATE THE MOTOR IN EVERY LOOP OF THE COMMAND**

jordan-powers replied:
Also, having the motors start moving when the button is pushed, then stop moving when the button is pushed again is really non-intuitive. Are you sure this what your drivers would prefer? It'd be better if they moved while the button was held down, and stopped when the button was released.

jordan-powers replied:
Also, the "move arms down or up" logic should be done in the control wrapper. This should just get the value from the controller, and write it to the arms.

So in DriveController.java, replace `getMoveArmsUp()` and `getMoveArmsDown()` with `getArmsSpeed()`.
In ArmAngle.java, see my comment about `setSpeed()`.

Then in ArmAngleCommand.java, replace `execute()` with 
```
protected void execute() {
    Robot.armAngle.setSpeed(Robot.controller.getArmsSpeed);
}
```

jordan-powers replied:
Then if you want to keep the "toggle" button, in F310Controller.java, implement getArmsSpeed() as:
```
private static final double ARM_UP_SPEED = 0.5;
private static final double ARM_DOWN_SPEED = -0.5;
private enum ArmState {MOVING_UP, MOVING_DOWN, STOPPED};
private ArmState currentState = ArmState.STOPPED;

@ Override
public double getArmsSpeed() {
    if(logitechf310.getXPressed()) {
        if(currentState == ArmState.STOPPED) {
            currentState = ArmState.MOVING_UP;
        } else {
            currentState = ArmState.STOPPED);
        }
    }
    if(logitechf310.getYPressed()) {
        if(currentState == ArmState.STOPPED) {
            currentState = ArmState.MOVING_DOWN;
        } else {
            currentState = ArmState.STOPPED);
        }
    }
    if(currentState == ArmState.MOVING_UP) {
        return ARM_UP_SPEED;
    } else if (currentState == ArmState.MOVING_DOWN) {
        return ARM_DOWN_SPEED;
    } else {
        return 0;
    }
}

jordan-powers replied:
Or if you want to switch to the vastly superior "hold button", do:
```
private static final double ARM_UP_SPEED = 0.5;
private static final double ARM_DOWN_SPEED = -0.5;
@ Override
public double getArmsSpeed() {
    if(logitechf310.getX()) {
        return ARM_UP_SPEED;
    } else if (logitechf310.getY()) {
        return ARM_DOWN_SPEED;
    } else {
        return 0;
    }
}
```

jordan-powers commented on src/main/java/frc/robot/commands/EngageSolenoidCommand.java:
```diff
@@ -0,0 +1,57 @@
+
+  // Called repeatedly when this Command is scheduled to run
+  @Override
+  protected void execute() {
```
This should be an [InstantCommand](http://first.wpi.edu/FRC/roborio/release/docs/java/edu/wpi/first/wpilibj/command/InstantCommand.html)
Remove the `if(controller.getStartLift())` check and instead make it so that this command is only ever run when you want it to

petervdonovan replied:
Will decided not to do that because it takes time to retract the climbing mechanism, so he wanted EngageSolenoidCommand finish() method to return true only after we think the solenoid should be done moving, based on a timer. The same is true of LiftRetractFront. The goal is to make it sequential so that the next command is blocked until the cylinders are done moving. Is that okay?

jordan-powers replied:
Yeah, just put nothing in execute and use a timer for isFinished(). But you still shouldn't check `if(controller.getStartLift())`

jordan-powers commented on src/main/java/frc/robot/commands/ExampleCommandGroup.java:
```diff
@@ -0,0 +1,34 @@
+/*----------------------------------------------------------------------------*/
```
Why does this file exist?

jordan-powers commented:
If you're unhappy with code, don't close the pull request. Leave a comment that you'd like updates, and any new commits will be included in the pull request

jordan-powers commented on src/main/java/frc/robot/commands/LiftCommandGroup.java:
```diff
@@ -0,0 +1,44 @@
+private final double stall2 = 2.0;
+  public LiftCommandGroup() {
+    addSequential(new EngageSolenoidCommand());
+    Timer.delay(stall);
```
NONONONONONONONONO

jordan-powers replied:
**NEVER** use  `Timer.delay()`. This will cause the entire robot program to stop running for 3 seconds!!

jordan-powers replied:
This won't even work the way you want it to. You're IN THE CONSTRUCTOR. You're not executing the command, you're just building it.
This method is run when you call `new LiftCommandGroup()`, not when the command is run!

jordan-powers replied:
Did you even read the docs?
https://wpilib.screenstepslive.com/s/currentCS/m/java/l/599738-creating-groups-of-commands
Do you see any `Timer.delay()`s in there?

jordan-powers commented on src/main/java/frc/robot/commands/LiftCommandGroup.java:
```diff
@@ -0,0 +1,44 @@
+private final double stall2 = 2.0;
+  public LiftCommandGroup() {
+    addSequential(new EngageSolenoidCommand());
+    Timer.delay(stall);
```
You don't understand how a command works.
A command has 3 important methods: `initialize()`, `execute()`, and `isFinished()`. A command, when started, will run `initialize()` once. It will then repeatedly run `execute()` until `isFinished()` becomes true. When `isFinished()` becomes true, the command will stop running

The point of a command group is that you want to run some commands all in a row. What the command group will do is it will run the first command in its internal list of commands, then wait for it to stop running (which it will do when `isFinished()` becomes true). Once that command finishes, it runs the next command in its list.

Now the question is how to add commands to the command group's list, and the answer is by using `addSequential()` and `addParallel()`. The list starts empty. Then you use `addSequential()` to add a command to the end of that empty list. Then you use `addSequential()` to add another command to the end of the list, so it will run after the previous command. If you want a command to run at the same time as a previous command, you use `addParallel()`.

I think an example is needed.
Let's say we want the robot to drive forward 1 meter. Then you want it to drive forward 2 meters while moving the arms up. Then you want it to shoot what's being held.
That would look something like this:
```
addSequential(new Drive(1));             // This will run first
addSequential(new Drive(2));             // This will run after the previous command finishes
addParallel(new SetArmsAngle(90));       // This will run at the same time as the previous command
addSequential(new Shoot());              // This will run after both of the previous commands finish
```

jordan-powers commented on src/main/java/frc/robot/commands/LiftDriveForward.java:
```diff
@@ -0,0 +1,49 @@
+
+  // Make this return true when this Command no longer needs to run execute()
+  @Override
+  protected boolean isFinished() {
```
There are a couple of ways to implement the isFinished() condition.
You can have a sensor on the subsystem, and you return true when the subsystem is where you want it.
Another option is to use a [Timer](http://first.wpi.edu/FRC/roborio/release/docs/java/edu/wpi/first/wpilibj/Timer.html). `timer.reset()` your timer in the `initialize()` method, then return `timer.hasPeriodPassed(DRIVE_FORWARD_PERIOD)`

petervdonovan replied:
We thought about using a limit switch, but putting one on the front climbing mechanism is hard because we don't already have wires going down there (no motors). Will and I decided to use a Timer

jordan-powers replied:
Is the mechanism hitting a physical stop? You can detect the current spike when the motors stall out and use that as your sensor

petervdonovan replied:
The gear ratio is very high -- 252:1. I am concerned that the wheels will just start slipping before a spike in current is detected that is large enough to be separated out from the noise

seg9585 replied:
Can the operator hold a button until the end condition is reached?

jordan-powers commented on src/main/java/frc/robot/commands/LiftSystemCommand.java:
```diff
@@ -8,6 +8,7 @@
 package frc.robot.commands;
```
So, what? This class doesn't do anything? Then why do you have it? To make your commands package more cluttered and harder to use?

jordan-powers commented on src/main/java/frc/robot/commands/LiftSystemEnd.java:
```diff
@@ -0,0 +1,47 @@
+  // Called repeatedly when this Command is scheduled to run
+  @Override
+  protected void execute() {
+    Robot.liftSystem.endLiftSequence(); 
```
Again, this should be an [InstantCommand](http://first.wpi.edu/FRC/roborio/release/docs/java/edu/wpi/first/wpilibj/command/InstantCommand.html)

jordan-powers commented on src/main/java/frc/robot/commands/LiftSystemEnd.java:
```diff
@@ -0,0 +1,47 @@
+import edu.wpi.first.wpilibj.command.Command;
+import frc.robot.Robot;
+
+public class LiftSystemEnd extends Command {
```
`LiftSystemEnd` is too specific of a name. Maybe call it something like `ResetLiftSystem` or `StopLiftSystem`?

jordan-powers commented on src/main/java/frc/robot/subsystems/LiftSystem.java:
```diff
@@ -17,47 +18,43 @@
+    front.set(Value.kReverse);
+    
+  }
+  public void endLiftSequence(){
```
This is a bad name. Think about what this method is doing.
Maybe call it `resetLift()` or `stopLift()`

jordan-powers commented on src/main/java/frc/robot/commands/WheelArmCommand.java:
```diff
@@ -27,15 +27,15 @@ protected void execute() {
     DriveController controller = Robot.controller;
     if(Robot.wheelArm.isInwards()||Robot.wheelArm.isOutwards()){
```
**ALWAYS UPDATE THE MOTOR IN EVERY LOOP OF THE COMMAND
ALWAYS, ALWAYS UPDATE THE MOTOR IN EVERY LOOP OF THE COMMAND
IF YOU WANT THE ROBOT TO WORK PROPERLY, UPDATE THE MOTOR IN EVERY LOOP OF THE COMMAND**
**_UPDATE THE MOTOR IN EVERY LOOP OF THE COMMAND_**
**UPDATE THE MOTOR IN EVERY LOOP OF THE COMMAND**

jordan-powers commented on src/main/java/frc/robot/commands/EngageSolenoidCommand.java:
```diff
@@ -0,0 +1,57 @@
+import edu.wpi.first.wpilibj.Timer;
+import frc.robot.controls.DriveController;
+
+public class EngageSolenoidCommand extends Command {
```
Be more descriptive, include the name of the subsystem
`EngageLiftSolenoidCommand`

jordan-powers commented on src/main/java/frc/robot/controls/F310Controller.java:
```diff
@@ -19,70 +19,72 @@
     @Override
     public double getMoveRequest(){
-        double speed = controller1.getY(Hand.kRight);
+        double speed = controller1.getRawAxis(1);
```
WHY? Why would you replace the easy to read, easy to understand, and easy to debug `getY()` with the arbitrary, easy to screw up `getRawAxis(1)`??

petervdonovan replied:
Juliana and Will said that the LogitechF310 class didn't work. I don't know anything more specific

jordan-powers commented on src/main/java/frc/robot/controls/F310Controller.java:
```diff
@@ -19,70 +19,72 @@
-    public boolean getToggleArmsUp(){
-        return controller1.getYButtonPressed();
+    
+    public boolean getMoveArmsUp(){
```
You lost the @ Override

jordan-powers commented on src/main/java/frc/robot/subsystems/ArmAngle.java:
```diff
@@ -8,14 +8,35 @@
+    super("Arm Angle");
+    addChild("Angle Motor", angle);
+  }
+  public void spinUpwards(){
```
You don't need three methods `spinUpwards()`, `spinDownwards()`, and `stopSpin()`. That's putting too much of the logic in the subsystem. There should only be one method `setSpeed(double speed)` that sets the speed of the arms. And yes, that does just make the subsystem into a seemingly unnecessary wrapper around a motor, but that's okay.

jordan-powers commented on src/main/java/frc/robot/subsystems/Drive.java:
```diff
@@ -10,7 +10,9 @@
 import edu.wpi.first.wpilibj.drive.DifferentialDrive;
+import edu.wpi.first.wpilibj.Spark;
 import frc.robot.RobotMap;
+import frc.robot.commands.DriveCommand;
```
Why are you importing this? You're not using it

jordan-powers commented on src/main/java/frc/robot/subsystems/HatchPusher.java:
```diff
@@ -20,32 +20,28 @@
-  private DoubleSolenoid right = RobotMap.hatchSolenoidRight;
+  
 
  private boolean out = false;
```
Unnecessary. Don't keep track of the state using a boolean. Just use `top.get()` and compare it to `Value.kForward`
```
public boolean isOut() {
    return top.get() == Value.kForward;
}
```

seg9585 replied:
better yet, just create a public boolean and set it when calling extend() or retract()

jordan-powers commented on src/main/java/frc/robot/subsystems/WheelArm.java:
```diff
@@ -25,21 +28,21 @@ public WheelArm(){
-    left.set(Value.kForward);//Check if Forwards means clockwise
-    right.set(Value.kReverse);
+    left.set(wheelArmSpeed);
+    right.set(-wheelArmSpeed);
```
This is bug-prone. Imagine you were changing something or adding to this class, and you forgot the negative in `right.set()`.  That would be pretty easy to forget.
What you should do is in the constructor, you should call `left.setInverted(false)` and `right.setInverted(true)`. Then create a [SpeedControllerGroup](http://first.wpi.edu/FRC/roborio/release/docs/java/edu/wpi/first/wpilibj/SpeedControllerGroup.html) called `wheelMotors`. Then create a method called `spin(double speed)` that just calls `wheelMotors.set(speed)`;

jordan-powers replied:
So the constructor would look something like this:
```
SpeedControllerGroup wheelMotors;
public WheelArm() {
    ...
    left.setInverted(false);
    right.setInverted(true);
    wheelMotors = new SpeedControllerGroup(left, right);
}
```

and spin would look something like this:
```
public spin(double speed) {
    wheelMotors.set(speed);
}
```

jordan-powers reviewed (changes requested):
Please actually read my comments, respond to them, and fix them. Reviews are a two-person communication, I can't always know what you're thinking so it helps if you explain.

petervdonovan committed:
Update EngageSolenoidCommand.java

petervdonovan committed:
Delete EngageSolenoidCommand.java

petervdonovan committed:
Create EngageLiftSolenoidCommand

petervdonovan committed:
Update LiftCommandGroup.java

Jason7042 committed:
"Fixed" Delays for Lift System

Jason7042 committed:
"Fixed" Lift System Part 2

jordan-powers commented on src/main/java/frc/robot/commands/EngageLiftSolenoidCommand:
```diff
@@ -12,11 +12,16 @@
 import edu.wpi.first.wpilibj.Timer;
 import frc.robot.controls.DriveController;
 
+<<<<<<< HEAD:src/main/java/frc/robot/commands/EngageSolenoidCommand.java
```
You messed up a merge conflict

jordan-powers committed:
Fixed merge conflict

petervdonovan committed:
Changed to instantCommand

petervdonovan committed:
Update WheelArm.java

Jason7042 committed:
"Fixed" Lift System Part 2

Jason7042 committed:
Merge branch 'CommandGroupTest' of https://github.com/poly-rabbotics/2019-Deep-Space into CommandGroupTest

petervdonovan committed:
Update F310Controller.java

petervdonovan committed:
Remove unnecessary imports

petervdonovan committed:
Fixed typo

petervdonovan committed:
Created a SpeedControllerGroup

niclop03 committed:
Camera Updates

niclop03 committed:
Merge branch 'CommandGroupTest' of https://github.com/poly-rabbotics/2019-Deep-Space into CommandGroupTest

niclop03 committed:
Moved Pixi2 Vision Code In From Shockwave4546

niclop03 committed:
pixy2 camera code

markpowers committed:
Fixing ArmAngle and Lift

markpowers committed:
Fixed ArmAngle and Lift(part 2)

niclop03 committed:
Update LiftCommandGroup.java

niclop03 committed:
Deleted LiftSystem End

niclop03 committed:
Yes

PolyRabbotics committed:
Move logic into the subsystem

PolyRabbotics committed:
Minimize logic in the command

It seems like ideally all the logic should be in the DriveController, but I wasn't sure that's what we wanted so I just thought I would make it not be as badly split between here and the command

PolyRabbotics committed:
Update WheelArmCommand.java

PolyRabbotics committed:
Update WheelArm.java

PolyRabbotics committed:
Update WheelArmCommand.java

PolyRabbotics committed:
Update WheelArmCommand.java

PolyRabbotics committed:
Update WheelArm.java

Bobman31 committed:
Revert "Yes"

This reverts commit 0d5d00de2f7da63dd6a2cdc43041090ef4c2d906.

Bobman31 committed:
Sure

seg9585 commented on src/main/java/frc/robot/Dashboard.java:
```diff
@@ -0,0 +1,54 @@
+
+    public static Dashboard getInstance() {
+        if (instance == null)
+            instance = new Dashboard();
```
Why are you constructing a new Dashboard here in a getInstance() method?  SmartDashboard is static anyway, just use it statically

jordan-powers replied:
In fact, what's even the point of this class?

seg9585 commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -44,7 +64,49 @@
+    // hatchCamera = CameraServer.getInstance().startAutomaticCapture("hatch",0);
+    // hatchCamera.setResolution(640, 480);
+     portCamera = CameraServer.getInstance().startAutomaticCapture("port",1);
+     portCamera.setResolution(1280, 960);
```
In my experience, you need to set camera parameters BEFORE starting automatic Capture, or the settings won't take effect.

seg9585 replied:
Yes I realize this is currently commented out.  Just an observation if you choose to bring it back

seg9585 commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -44,7 +64,49 @@
+    }).start();
+    */
+		Cameras.setup(); // Setup and Connection to Pixy2 and Microsoft Camera
+		SmartDashboard.putNumber("Iteration", myCounter);
```
robotInit() only runs 1 time, what is myCounter for?

seg9585 commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -44,7 +64,49 @@
+			 * Multiple navX-model devices on a single robot are supported.
+			 ************************************************************************/
+            //ahrs = new AHRS(SerialPort.Port.kUSB1);
+            ahrs = new AHRS(SerialPort.Port.kMXP, SerialDataType.kProcessedData, (byte)50);
```
Is this line causing runtime errors?  If you try-catch this and it fails, the code will not try again in robotInit.  This error may occur if the navX is not installed or working properly, I presume

seg9585 commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -120,6 +262,8 @@ public void teleopInit() {
     new DriveCommand().start();
```
These should not be called every iteration of robotPeriodic().   You should be using the Scheduler   Scheduler.getInstance.run();

seg9585 commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -120,6 +262,8 @@ public void teleopInit() {
     new DriveCommand().start();
```
Are you planning to drive through Sandstorm?  If so, these will need to be run in autonomousInit().     teleopInit() will not start until the end of sandstorm

jordan-powers replied:
The file does include the [auto methods](https://github.com/poly-rabbotics/2019-Deep-Space/blob/f21965d3cfbda66b14802de1d0dba7259b6277d9/src/main/java/frc/robot/Robot.java), GitHub is only showing a diff not the whole file. `autonomousInit()` does need to have a `DriveCommand` though

seg9585 commented on src/main/java/frc/robot/RobotMap.java:
```diff
@@ -28,26 +29,22 @@
   // number and the module. For example you with a rangefinder:
   // public static int rangefinderPort = 1;
   // public static int rangefinderModule = 1;
-  public static final Spark leftFront = new Spark(0);// TODO: Set to actual port numbers
```
In CommandBase, the objects are created within their subsystems, not in RobotMap.   RobotMap should only be defining static final integers specifying ports

jordan-powers replied:
How are you going to make sure there's only ever one object associated with a port then? Also, in the [FRC Docs](https://wpilib.screenstepslive.com/s/currentCS/m/java/l/599735-simple-subsystems) they use `Victor motor = RobotMap.clawMotor;`

seg9585 replied:
Ok, just a different way to do it I guess.   I've always seen this in subsystem:      Victor motor = new Victor(RobotMap.clawMotor);

jordan-powers commented on README.md:
```diff
@@ -1,4 +0,0 @@
-# 2019-Deep-Space
```
Why delete?

jordan-powers commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -57,6 +119,86 @@ public void robotInit() {
+            /* Connectivity Debugging Support                                           */
+            SmartDashboard.putNumber(   "IMU_Byte_Count",       ahrs.getByteCount());
+            SmartDashboard.putNumber(   "IMU_Update_Count",     ahrs.getUpdateCount());
+		Cameras.run();
```
This is a **lot** of stuff to be printing to the console. Might make it hard to see important stuff on the console

jordan-powers commented on src/main/java/frc/robot/commands/EngageLiftSolenoidCommand.java:
```diff
@@ -0,0 +1,60 @@
+  // Make this return true when this Command no longer needs to run execute()
+  @Override
+  protected boolean isFinished() {
+    return (t.get()>=SOLENOID_DELAY);
```
Timer has a method `hasPeriodPassed()` http://first.wpi.edu/FRC/roborio/release/docs/java/edu/wpi/first/wpilibj/Timer.html#hasPeriodPassed(double)

jordan-powers commented on src/main/java/frc/robot/commands/LiftCommandGroup.java:
```diff
@@ -0,0 +1,41 @@
+/*----------------------------------------------------------------------------*/
```
What's the point of this class if it's not doing anything?

jordan-powers commented on src/main/java/frc/robot/commands/LiftDriveForwardBack.java:
```diff
@@ -0,0 +1,44 @@
+  // Make this return true when this Command no longer needs to run execute()
+  @Override
+  protected boolean isFinished() {
+    return (t.get()>=DRIVE_DELAY);
```
http://first.wpi.edu/FRC/roborio/release/docs/java/edu/wpi/first/wpilibj/Timer.html#hasPeriodPassed(double)

jordan-powers commented on src/main/java/frc/robot/commands/LiftDriveForwardFront.java:
```diff
@@ -0,0 +1,55 @@
+  // Make this return true when this Command no longer needs to run execute()
+  @Override
+  protected boolean isFinished() {
+    return (t.get()>=DRIVE_DELAY);
```
http://first.wpi.edu/FRC/roborio/release/docs/java/edu/wpi/first/wpilibj/Timer.html#hasPeriodPassed(double)

jordan-powers commented on src/main/java/frc/robot/commands/LiftRetractFront.java:
```diff
@@ -0,0 +1,55 @@
+  // Make this return true when this Command no longer needs to run execute()
+  @Override
+  protected boolean isFinished() {
+    return (t.get()>=RETRACT_DELAY);
```
http://first.wpi.edu/FRC/roborio/release/docs/java/edu/wpi/first/wpilibj/Timer.html#hasPeriodPassed(double)

jordan-powers commented on src/main/java/frc/robot/controls/DriveController.java:
```diff
@@ -11,12 +11,15 @@
  */
 public interface DriveController {
-    public double getMoveRequest();
+    Object getArmsSpeed = null;
```
What is this? This is impossible, you can't have an instance variable on an interface

jordan-powers commented on src/main/java/frc/robot/subsystems/HatchPusher.java:
```diff
@@ -20,36 +20,30 @@
 
   public boolean isOut(){
-    return out;
+    return top.get()==Value.kForward;
```
You replaced the boolean with a comparison, but you didn't delete the boolean from the class

jordan-powers commented on src/main/java/frc/robot/subsystems/vision/TargetBall.java:
```diff
@@ -0,0 +1,44 @@
+//import io.github.pseudoresonance.pixy2api.Pixy2CCC;
+import io.github.pseudoresonance.pixy2api.Pixy2CCC.Block;
+
+public class TargetBall {
```
How much ball following is in this year's game? Maybe don't merge test code to the main robot code branch?

jordan-powers commented on src/main/java/org/usfirst/frc/team4999/controllers/LogitechF310.java:
```diff
@@ -125,6 +125,10 @@ public boolean getBackButtonPressed() {
 	}
+
+	public double getArmsSpeed() {
+		return 0;
```
So your arms are never  going to move?
Come on, I basically wrote two implementations of this method for you and you can't be bothered to chose one and paste it in?

jordan-powers committed:
Made the changes which weren't being made

niclop03 committed:
Update build.gradle

niclop03 committed:
Fixed issues with Drive
